### PR TITLE
[SYCL] Fix linking of kernel-bundles.

### PR DESCRIPTION
**Pull request by cperkinsintel**
when `fsycl-device-code-split=per_kernel` is used in conjunction with multiple kernels, then explicit `sycl::link(sycl::compile(my-kernel-bundle))` will fail, because we link all device images in a kernel together. Any shared top level references will have conflicting symbols. But, the device-images are independent, there is presently no known situation where they need to be linked together. So this PR works around this limitation by linking them separately. This may have to be revisited once sharedlibrary style linking is supported, but that is likely true whether this bug is fixed or not.

Signed-off-by: Chris Perkins <chris.perkins@intel.com>

### CODE REVIEW

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 1]
```diff
@@ -192,7 +192,6 @@ class kernel_bundle_impl {
     // TODO: Unify with c'tor for sycl::comile and sycl::build by calling
```
Comment (romanovvlad) [inline sycl/source/detail/kernel_bundle_impl.hpp:192]: Can we add a test to sycl/unittests/SYCL2020 ?

Comment (cperkinsintel) [reply sycl/source/detail/kernel_bundle_impl.hpp:192]: I have a lit test for the multiple kernel stuff here: https://github.com/intel/llvm-test-suite/pull/440
Is that what you are asking?

**Comment by gmlueck** [comment]
I think it would be worth adding a comment explaining the limitations.  Something like:

> The loop below just links each device image separately, not linking any two device images together.  This is correct so long as each device image has no unresolved symbols.  That's the case when device images are created from generic SYCL APIs.  There's no way in generic SYCL to create a kernel which references an undefined symbol.  If we decide in the future to allow a backend interop API to create a "sycl::kernel_bundle" that references undefined symbols, then the logic in this loop will need to be changed.

**Comment by cperkinsintel** [comment]
The matching test for this is here: https://github.com/intel/llvm-test-suite/pull/440
And here is the run of that and this PR together and passing: http://icl-jenkins.sc.intel.com:8080/job/SYCL_CI/job/intel/job/Lin/job/LLVM_Test_Suite/7433/

**Comment by cperkinsintel** [comment]
@gmlueck  - thanks. I inserted your comment.

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 2]
```diff
@@ -192,6 +192,7 @@ class kernel_bundle_impl {
...
     // sycl::join on vector of kernel_bundles
 
+    // The loop below just links each device image separately, not linking any two device images together. This is correct so long as each device image has no unresolved symbols. That's the case when device images are created from generic SYCL APIs. There's no way in generic SYCL to create a kernel which references an undefined symbol. If we decide in the future to allow a backend interop API to create a "sycl::kernel_bundle" that references undefined symbols, then the logic in this loop will need to be changed.
```
Comment (gmlueck) [inline sycl/source/detail/kernel_bundle_impl.hpp:195]: I presume we have some line length limit in our source base, so this should be split into multiple lines?

Comment (cperkinsintel) [reply sycl/source/detail/kernel_bundle_impl.hpp]: No, we rely on clang-format and our dev team to remember to ADD the clang-format modification back into their commits.  I ran clang-format, but forgot to adjust the commit with it. Just bumped it.  Thanks.

**Review by gmlueck — APPROVED** [review]

**Review by romanovvlad — APPROVED** [review]

File: sycl/source/detail/kernel_bundle_impl.hpp (hunk 1 above)
Comment (romanovvlad) [reply sycl/source/detail/kernel_bundle_impl.hpp:192]: No, I was asking about adding a unit/component test in `sycl/unittests/SYCL2020` directory of this repo. This test should be able to validate this behavior without any backend available.
Let's do it in a separate PR?

**Pull request merged**